File: 653-two-sum-iv-input-is-a-bst/653-two-sum-iv-input-is-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    void inorder(TreeNode* root,vector<int>& in){
        if(root == NULL) return;
        
        if(root->left) inorder(root->left,in);
        in.push_back(root->val);
        if(root->right) inorder(root->right,in);
    }
public:
    bool findTarget(TreeNode* root, int k) {
        vector<int> in;
        inorder(root,in);
        //for(auto it: in) cout << it << " ";
        
        unordered_map<int,int> mpp;
        
        for(int i=0;i<in.size();i++){
            if(mpp[k-in[i]] == 1) return true;
            mpp[in[i]] = 1;
        }
        
        return false;
    }
};

File: 653-two-sum-iv-input-is-a-bst/653-two-sum-iv-input-is-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    

            
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        order = []
        
        
        def inorder(root : Optional[TreeNode], order : List[int]):
            if not root : 
                return

            if root.left :
                inorder(root.left,order)
            order.append(root.val)

            if root.right:
                inorder(root.right,order)
        
        inorder(root,order)
        temp = {}
        
        for i,num in enumerate(order):
            if k-num in temp:
                return True
            
            temp[num] = 1
        
        return False
        
        